testing/asserts.ts

```diff
@@ -749,11 +749,19 @@ export async function assertRejects<E extends Error = Error>(
     msg = msgIncludesOrMsg;
   }
   let doesThrow = false;
+  let isPromiseReturned = false;
   try {
-    await fn();
+    const possiblePromise = fn();
+    if (possiblePromise instanceof Promise) {
+      isPromiseReturned = true;
+      await possiblePromise;
+    }
   } catch (error) {
+    if (!isPromiseReturned) {
+      throw new AssertionError("Function throws when expected to reject.");
+    }
     if (error instanceof Error === false) {
-      throw new AssertionError("A non-Error object was thrown or rejected.");
+      throw new AssertionError("A non-Error object was rejected.");
```

Comment:
Only question left is whether or not these 2 errors should have msg appended to them like the "Expected function to reject" message at the end of this function. I don't have any strong opinions on that. If we do make that change, I think `assertThrows` AssertionError messaging should be updated too.